### ISSUE TITLE
Softcapping for lm head logits?

### ISSUE THREAD

**Issue by kallewoof** [tdrussell/qlora-pipe#18]
The Unsloth people made a [blog post](https://unsloth.ai/blog/gemma2) about Gemma2, and they noted:

> it's a must to turn on softcapping for the lm head logits, whilst the attention softcapping is generally recommended as well. If you do not turn it on for the lm head, your training loss will be incorrect

I wonder if this is something that needs to be fixed in qlora-pipe somehow. I haven't really wrapped my head around what they're suggesting yet, but figured I'd put up an issue before it escapes my head.

**Comment by tdrussell** [comment]
All the logit softcapping should already be handled. For the attention softcapping, the pipeline model wraps the entire original decoder layer so that is taken care of. And for the lm_head, I made an lm_head wrapper class specifically for Gemma2 called Gemma2LmHeadPipe which does the logit softcapping (just copied from the Transformers implementation).

But Gemma2 has had many implementation issues across backends so I'm not 100% convinced everything is correct. If something looks wrong or behaves weird with training just let me know.

**Comment by kallewoof** [comment]
Done some training and it seems fine. Thanks!

**Comment by jukofyork** [comment]
Just noticed unsloth has some extra parameters for `Cohere` and `Gemma2` now:

https://github.com/unslothai/unsloth/blob/main/unsloth/kernels/cross_entropy_loss.py

```python
    # Go logit scaling for Cohere: t * x
    if DO_LOGIT_SCALING: logits = LOGIT_SCALE * logits
    # Do logit softcapping for Gemma 2: t * tanh(1/t * x)
    if DO_SOFTCAPPING:   logits = SOFTCAP * triton_tanh(logits / SOFTCAP)
 ```

The LOGIT_SCALE seems to be fixed at `5/6` and set here:

```
  "logit_scale": 0.8333333333333334,
```

Apologies if you've already seen this :)
